### PR TITLE
Fixes #20804 - Support for deprecating commands

### PR DESCRIPTION
**Pull request by rahulbajaj0509**

### CODE REVIEW

**Comment by rahulbajaj0509** [comment]
@tstrachota hey, i have tried doing this, it is not working as expected, can you give some input to it ?

**Comment by tstrachota** [comment]
@rahulbajaj0509 what exactly does not work for you? I gave it a try and it seems to work:

```ruby
    HammerCLI::MainCommand.lazy_subcommand('architecture', _("Manipulate architectures."),
      'HammerCLIForeman::Architecture', 'hammer_cli_foreman/architecture',
      true
    )
```
...and the architecture is gone from `hammer -h`.

We need to extend also `subcommand` though. Not all commands are loaded lazily in hammer.

**Comment by rahulbajaj0509** [comment]
@tstrachota actually the lazy_subcommand that i was passing was something like : 

```
HammerCLI::MainCommand.lazy_subcommand('add-override-value', _("Override the param"),
      'HammerCLIForeman::SmartClassParameter::AddOverrrideValueCommand',
       'haammer_cli_foreman/smart_class_parameter',
       true
    )
```

So, that is the reason it was not working for me.

File: lib/hammer_cli/subcommand.rb [hunk 1]
```diff
@@ -44,13 +66,29 @@ def remove_subcommand(name)
...
+      end
+
+      def lazy_subcommand!(name, description, subcommand_class_name, path, options = {})
+        remove_subcommand(name)
+        self.lazy_subcommand(name, description, subcommand_class_name, path, options)
+        logger.info "subcommand #{name} (#{subcommand_class_name}) was created."
+      end
+
+      def define_subcommand(subcommand_class, definition, &block)
         existing = find_subcommand(name)
```
Comment (kgaikwad) [inline lib/hammer_cli/subcommand.rb]: @rahulbajaj0509 ,

Does `name` accessible inside this method? Everytime you will get nil value for existing.

Comment (rahulbajaj0509) [reply lib/hammer_cli/subcommand.rb]: Thanks @kgaikwad :)

**Comment by rahulbajaj0509** [comment]
[test]

**Comment by rahulbajaj0509** [comment]
@tstrachota i am not sure why `checks are failing` here, there are no tests failing though.

**Comment by tstrachota** [comment]
[test]

**Comment by tstrachota** [comment]
@rahulbajaj0509 The code looks good now. I missed the failing test and even merged it but then I realized the test failure is related and had to revert it. There's problem with Clamp on ruby 2.0 where we have it pinned to < 1.1.0. Clamp dropped 2.0 support since then.

This is the commit that we're missing:
https://github.com/mdub/clamp/commit/d0c4288a543b65f64804f64f92ed5e0cee47de85

I haven't found any straightforward and clean way of adding the compatibility yet. @rahulbajaj0509 @mbacovsky any ideas?

**Comment by tstrachota** [comment]
[test]

**Comment by tstrachota** [comment]
:+1: thanks @rahulbajaj0509

File: lib/hammer_cli/subcommand.rb [hunk 2]
```diff
@@ -59,23 +97,10 @@ def subcommand(name, description, subcommand_class = self, &block)
             :new_class => subcommand_class
           }
         end
-        super
+        subcommand_class = Class.new(subcommand_class, &block) if block
+        declare_subcommand_parameters unless has_subcommands?
```
Comment (ekohl) [inline lib/hammer_cli/subcommand.rb]: This line breaks systest: https://ci.theforeman.org/job/systest_foreman/13810/console

```
not ok 16 check smart proxy is registered and Hammer runs
# (in test file /usr/bin/fb-install-foreman.bats, line 167)
#   `count=$(hammer $(tHammerCredentials) --csv proxy list | wc -l)' failed
# /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/subcommand.rb:101:in `define_subcommand': undefined local variable or method `declare_subcommand_parameters' for HammerCLI::DefaultsCommand:Class (NameError)
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/subcommand.rb:77:in `subcommand'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/abstract.rb:212:in `block in autoload_subcommands'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/abstract.rb:211:in `each'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/abstract.rb:211:in `autoload_subcommands'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/defaults_commands.rb:158:in `<class:DefaultsCommand>'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/defaults_commands.rb:22:in `<module:HammerCLI>'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/defaults_commands.rb:3:in `<top (required)>'
# 	from /opt/rh/rh-ruby22/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:54:in `require'
# 	from /opt/rh/rh-ruby22/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:54:in `require'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/defaults.rb:1:in `<top (required)>'
# 	from /opt/rh/rh-ruby22/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:54:in `require'
# 	from /opt/rh/rh-ruby22/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:54:in `require'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli/context.rb:1:in `<top (required)>'
# 	from /opt/rh/rh-ruby22/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:54:in `require'
# 	from /opt/rh/rh-ruby22/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:54:in `require'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/lib/hammer_cli.rb:17:in `<top (required)>'
# 	from /opt/rh/rh-ruby22/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:54:in `require'
# 	from /opt/rh/rh-ruby22/root/usr/share/rubygems/rubygems/core_ext/kernel_require.rb:54:in `require'
# 	from /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.12.pre.develop/bin/hammer:117:in `<top (required)>'
# 	from /usr/bin/hammer:23:in `load'
# 	from /usr/bin/hammer:23:in `<main>'
```

Comment (mbacovsky) [reply lib/hammer_cli/subcommand.rb]: @ekohl this change require new version of clamp. The packaging PRs are https://github.com/theforeman/foreman-packaging/pull/1872 (RPM) and https://github.com/theforeman/foreman-packaging/pull/1874 (DEB).

**Pull request merged**